### ISSUE TITLE
9. Ensemble_RandomForest

### ISSUE THREAD

**Issue by hih0124** [hih0124/lecture_python-II#44]
"""
Random Forest Ensemble model
###  배깅(Bagging)
- 병렬학습 : 각 모델의 결과 를 조합하여 투표 결정
- randomForest
### 부스팅(Boosting) 
- 순차학습 : 현재 모델 가중치 -> 다음 모델 전달
- XGboost
"""

from sklearn.ensemble import RandomForestClassifier # model
from sklearn.datasets import load_wine  # dataset
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report# 모델 평가

### 1. dataset load
wine = load_wine()
wine.feature_names # X변수명
wine.target_names # y범주 이름 # array(['class_0', 'class_1', 'class_2'], dtype='<U7')

X = wine.data
y = wine.target

X.shape # (178, 13)

### 2. RF model
rf = RandomForestClassifier() # 객체 생성
'''
n_estimators : integer, optional (default=100) : 트리 수
criterion : string, optional (default="gini") or "entropy" : 중요변수 선정기준
max_depth : integer or None, optional (default=None) : 트리 깊이
min_samples_split : int, float, optional (default=2) : 노드 분할할때 사용되는 최소 샘플 수
min_samples_leaf : int, float, optional (default=1) : 단노드 분할 최소 샘플 수
max_features : int, float, string or None, optional (default="auto") : 최대 X변수 사용 수
n_jobs 
random_state
'''
import numpy as np
idx = np.random.choice(a = X.shape[0], size=int(X.shape[0]*0.7), replace=False) # a: 전체 데이터
len(idx) # 124
X_train = X[idx] #X[id,:]
y_train = y[idx]
model = rf.fit(X=X_train, y=y_train)

idx_test = [i for i in range(len(X)) if not i in idx]
len(idx_test)
X_test = X[idx_test]
y_test = y[idx_test]

y_pred = model.predict(X_test)
y_true = y_test

con_mat = confusion_matrix(y_true,y_pred)
acc = accuracy_score(y_true, y_pred)
report = classification_report(y_true, y_pred)

con_mat
'''
array([[23,  0,  0],
       [ 0, 22,  0],
       [ 0,  0,  9]], dtype=int64)
'''
acc # 1.0
print(report)
'''
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        23
           1       1.00      1.00      1.00        22
           2       1.00      1.00      1.00         9

    accuracy                           1.00        54
   macro avg       1.00      1.00      1.00        54
weighted avg       1.00      1.00      1.00        54
'''

### 중요변수
print('중요변수 :',model.feature_importances_)
'''
중요변수 : [0.09793838 0.02854261 0.01491171 0.0363228  0.02456062 0.05036014
 0.17818634 0.01298509 0.02239279 0.15899339 0.10675822 0.10742115
 0.16062677]
지니계수를 기준으로 13개 변수에 대한 중요변수를 이렇게 점수화해서 반환함
'''

### 중요변수 시각화
import matplotlib.pyplot as plt
x_size = X.shape[1]
plt.barh(range(x_size), model.feature_importances_) # (y,x) 가로막대 차트
plt.yticks(range(x_size), wine.feature_names)
plt.xlabel('importance')
plt.show()
